server/embedded/src/org/labkey/embedded/EmbeddedExtractor.java

```diff
@@ -0,0 +1,149 @@
+package org.labkey.embedded;
+
+import org.labkey.bootstrap.ConfigException;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+public class EmbeddedExtractor
+{
+    private static final int BUFFER_SIZE = 1024 * 64;
+
+    private void extractExecutableJar(File jarFilePath, File destDirectory, boolean remotePipeline)
+    {
+        try
+        {
+            try (JarFile jar = new JarFile(jarFilePath))
+            {
+                boolean foundDistributionZip = false;
+                var entries = jar.entries();
+                while (entries.hasMoreElements())
+                {
+                    var entry = entries.nextElement();
+                    var entryName = entry.getName();
+
+                    if ("labkey/distribution.zip".equals(entryName))
+                    {
+                        foundDistributionZip = true;
+                        try (var distInputStream = jar.getInputStream(entry))
+                        {
+                            extractZip(distInputStream, destDirectory);
+                        }
+                    }
+                    if (remotePipeline)
+                    {
+                        if (entry.getName().contains("labkeyBootstrap") && entry.getName().toLowerCase().endsWith(".jar"))
+                        {
+                            try (var in = jar.getInputStream(entry))
+                            {
+                                extractFile(in, new File(destDirectory, "labkeyBootstrap.jar"));
+                            }
+                        }
+                        if (entry.getName().contains("tomcat-servlet-api") && entry.getName().toLowerCase().endsWith(".jar"))
+                        {
+                            File pipelineLib = new File(destDirectory, "pipeline-lib");
+                            if (!pipelineLib.exists())
+                            {
+                                if (!pipelineLib.mkdirs())
+                                {
+                                    throw new ConfigException("Failed to create directory " + pipelineLib + " Please check file system permissions");
+                                }
+                            }
+                            try (var in = jar.getInputStream(entry))
+                            {
+                                extractFile(in, new File(pipelineLib, "servletApi.jar"));
+                            }
+                        }
+                    }
+                }
+
+//                if (!foundDistributionZip)
+//                {
+//                    throw new ConfigException("Unable to find distribution zip required to run LabKey Server.");
+//                }
```

Comment:
Did you mean to make this non-fatal? I presume this is commented out for local testing?

Reply:
Indeed, it was to work on my local dev build. I restored it, though we can revisit as we decide exactly how/when to extract resources in upgrade scenarios.